### PR TITLE
[v2.0.0] Package the plugins

### PR DESCRIPTION
**Pull request by yoannmoinet**
### What and why?

We want to publish 2 packages and have them work as expected OOTB.

- `@datadog/webpack-plugin`
- `@datadog/esbuild-plugin`

### How?

We now bundle the published packages with rollup, using a shared configuration.
We had to refactor what was published and what not.
We've assessed the usage from web-ui using prerelease versions.

Fixes https://github.com/DataDog/build-plugin/issues/59

Tested it in https://github.com/DataDog/web-ui/pull/136308

### CODE REVIEW

**Comment by yoannmoinet** [comment]
Hello World

File: .eslintrc.js [hunk 1]
```diff
@@ -376,5 +375,16 @@ module.exports = {
...
         },
+        {
+            files: [
+                'rollup.config.mjs',
+                'packages/core/**/*',
+                'packages/webpack-plugin/**/*',
+                'packages/esbuild-plugin/**/*',
+            ],
+            rules: {
+                'import/no-extraneous-dependencies': 0,
```
Comment (Ayc0) [inline .eslintrc.js:386]:
```suggestion
                'import/no-extraneous-dependencies': 'error',
```

File: packages/core/src/types.ts [hunk 2]
```diff
@@ -244,7 +244,6 @@ export type GetPlugins<T> = (options: T) => UnpluginOptions[];
 export interface GetPluginsOptions {
     auth: {
         apiKey: string;
-        appKey: string;
```
Comment (Ayc0) [inline packages/core/src/types.ts:247]: Why did you remove the `appKey`?

File: packages/esbuild-plugin/package.json [hunk 3]
```diff
@@ -1,40 +1,71 @@
 {
     "name": "@datadog/esbuild-plugin",
     "packageManager": "yarn@4.0.2",
+    "version": "0.0.13-4",
     "license": "MIT",
     "author": "Datadog",
+    "description": "Datadog ESBuild Plugin.",
```
Comment (Ayc0) [inline packages/esbuild-plugin/package.json:7]: Nit: you may want to drop the `.` at the end of all your descriptions

File: packages/esbuild-plugin/package.json [hunk 4]
```diff
@@ -1,40 +1,71 @@
...
+    "repository": {
+        "type": "git",
+        "url": "https://github.com/DataDog/build-plugin",
+        "directory": "packages/esbuild-plugin"
+    },
     "main": "./dist/src/index.js",
+    "module": "./dist/src/index.mjs",
     "exports": {
         ".": "./src/index.ts",
         "./*": "./src/*.ts"
```
Comment (Ayc0) [inline packages/esbuild-plugin/package.json:25]: Shouldn't we drop the `./*` (as it's not overridden in the `publishConfig` which means that it won't work for built code)

File: packages/core/src/types.ts (hunk 2 above)
Comment (yoannmoinet) [reply packages/core/src/types.ts:247]: It was not used anywhere.

File: packages/esbuild-plugin/package.json [hunk 5]
```diff
@@ -1,40 +1,71 @@
...
+                "types": "./dist/src/index.d.ts"
+            }
         }
     },
     "files": [
         "dist"
     ],
     "scripts": {
-        "build": "yarn clean && tsc",
+        "build": "yarn clean && rollup --config rollup.config.mjs",
```
Comment (Ayc0) [inline packages/esbuild-plugin/package.json:43]: If the code is built by `rollup`, you way want to add `"moduleResolution": "bundler"` & `"module": "NodeNext"` in your TSConfig.json

File: packages/esbuild-plugin/package.json [hunk 6]
```diff
@@ -1,40 +1,71 @@
...
         "@dd/factory": "workspace:*",
-        "esbuild": "0.20.2"
+        "@dd/tools": "workspace:*",
+        "@rollup/plugin-babel": "6.0.4",
+        "@rollup/plugin-commonjs": "25.0.7",
+        "@rollup/plugin-json": "6.1.0",
+        "@rollup/plugin-node-resolve": "15.2.3",
+        "@types/babel__core": "^7",
+        "@types/babel__preset-env": "^7",
+        "esbuild": "0.20.2",
```
Comment (Ayc0) [inline packages/esbuild-plugin/package.json:61]: Note: you could also use `esbuild` as your main builder instead of having to install:
- rollup
- all the rollup plugins
- babel
- all the babel plugins

You can follow this guide for how to setup esbuild: https://esbuild.github.io/getting-started/#bundling-for-node

File: packages/plugins/telemetry/package.json [hunk 7]
```diff
@@ -28,11 +35,7 @@
         "esbuild": "0.20.2"
     },
     "peerDependencies": {
+        "esbuild": "*",
         "webpack": "*"
```
Comment (Ayc0) [inline packages/plugins/telemetry/package.json:39]: You should add min versions, like `"esbuild": ">=1.2.3"`

File: packages/tests/package.json [hunk 8]
```diff
@@ -14,16 +21,17 @@
         "test": "NODE_OPTIONS=\"--openssl-legacy-provider ${NODE_OPTIONS:-}\" jest --verbose"
     },
     "dependencies": {
-        "@datadog/build-plugins-core": "workspace:*",
-        "@datadog/esbuild-plugin": "workspace:*",
-        "@datadog/webpack-plugin": "workspace:*",
+        "@datadog/esbuild-plugin": "workspace:0.0.13-4",
+        "@datadog/webpack-plugin": "workspace:0.0.13-4",
```
Comment (Ayc0) [inline packages/tests/package.json:25]: To note: you shouldn't used pinned version, but `workspace:*`

yarn will use the version automatically of the other package on publish 
And if you publish them in order, it should be okay (this is what I do with my packages and it works amazingly good)

File: packages/webpack-plugin/package.json [hunk 9]
```diff
@@ -1,41 +1,71 @@
...
+        "rollup": "4.17.1",
+        "rollup-plugin-dts": "6.1.0",
+        "rollup-plugin-esbuild": "6.1.1",
+        "typescript": "5.4.3"
     },
     "peerDependencies": {
         "webpack": ">= 4.x < 6.x"
-    }
+    },
+    "stableVersion": "0.0.12-dev"
```
Comment (Ayc0) [inline packages/webpack-plugin/package.json:70]: what is `stableVersion`?

**Review by Ayc0 — APPROVED** [review]

File: .eslintrc.js (hunk 1 above)
Comment (yoannmoinet) [reply .eslintrc.js:386]: Fixed in https://github.com/DataDog/build-plugin/pull/67
In this case, you wanted it to be `'off'` right?

File: packages/esbuild-plugin/package.json (hunk 3 above)
Comment (yoannmoinet) [reply packages/esbuild-plugin/package.json:7]: Will do, but why?

File: packages/esbuild-plugin/package.json (hunk 4 above)
Comment (yoannmoinet) [reply packages/esbuild-plugin/package.json:25]: Good catch.

File: packages/esbuild-plugin/package.json (hunk 5 above)
Comment (yoannmoinet) [reply packages/esbuild-plugin/package.json:43]: Used `"preserve"` as discussed, except for the tests package.

File: packages/esbuild-plugin/package.json (hunk 6 above)
Comment (yoannmoinet) [reply packages/esbuild-plugin/package.json:61]: I'm fine with rollup right now, will change if we get to a point where it's necessary to change.

File: packages/plugins/telemetry/package.json (hunk 7 above)
Comment (yoannmoinet) [reply packages/plugins/telemetry/package.json:39]: Good call.

File: packages/esbuild-plugin/package.json (hunk 3 above)
Comment (Ayc0) [reply packages/esbuild-plugin/package.json:7]: Almost no-one uses `.` at the end of descriptions

File: packages/webpack-plugin/package.json (hunk 9 above)
Comment (Ayc0) [reply packages/webpack-plugin/package.json:70]: TIL
